src/api/MainSolver.cc

```diff
@@ -37,93 +19,131 @@ WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 
-void MainSolver::insertFormula(PTRef root)
+sstat MainSolver::getStatus() const { return status; }
+
+void MainSolver::insertFormula(PTRef fla)
 {
-    if (logic.getSortRef(root) != logic.getSort_bool()) {
-        throw OsmtApiException("Top-level assertion sort must be Bool, got " + logic.printSort(logic.getSortRef(root)));
+    if (logic.getSortRef(fla) != logic.getSort_bool()) {
+        throw OsmtApiException("Top-level assertion sort must be Bool, got " + logic.printSort(logic.getSortRef(fla)));
     }
-
-    root = logic.conjoinExtras(root);
-    root = IteHandler(logic, getPartitionManager().getNofPartitions()).rewrite(root);
+    // TODO: Move this to preprocessing of the formulas
+    fla = IteHandler(logic, getPartitionManager().getNofPartitions()).rewrite(fla);
 
     if (getConfig().produce_inter()) {
         // MB: Important for HiFrog! partition index is the index of the formula in an virtual array of inserted formulas,
         //     thus we need the old value of count. TODO: Find a good interface for this so it cannot be broken this easily
-        unsigned int partition_index = inserted_formulas_count++;
-        pmanager.assignTopLevelPartitionIndex(partition_index, root);
-        assert(pmanager.getPartitionIndex(root) != -1);
-    }
-    else {
-        ++inserted_formulas_count;
+        unsigned int partition_index = insertedFormulasCount++;
+        pmanager.assignTopLevelPartitionIndex(partition_index, fla);
+        assert(pmanager.getPartitionIndex(fla) != -1);
     }
+    else { ++insertedFormulasCount; }
 
-    PushFrame& lastFrame =  pfstore[frames.last()];
-    lastFrame.push(root);
-    lastFrame.root = PTRef_Undef;
-    // New formula has been added to the last frame. If the frame has been simplified before, we need to do it again
-    frames.setSimplifiedUntil(std::min(frames.getSimplifiedUntil(), frames.size() - 1));
+    frames.add(fla);
+    firstNotSimplifiedFrame = std::min(firstNotSimplifiedFrame, frames.frameCount() - 1);
 }
 
-sstat MainSolver::simplifyFormulas()
-{
+sstat MainSolver::simplifyFormulas() {
     status = s_Undef;
-
-    vec<PTRef> coll_f;
-    bool keepPartitionsSeparate = getConfig().produce_inter();
-    // Process (and simplify) not yet processed frames. Stop processing if solver is in UNSAT state already
-    for (std::size_t i = frames.getSimplifiedUntil(); i < frames.size() && status != s_False; i++) {
-        getTheory().simplify(frames.getFrameReferences(), pmanager, i);
-        frames.setSimplifiedUntil(i + 1);
-        const PushFrame & frame = pfstore[frames.getFrameReference(i)];
-
-        if (keepPartitionsSeparate) {
-            vec<PTRef> const & flas = frame.formulas;
-            if (flas.size() == 0 or std::all_of(flas.begin(), flas.end(), [&](PTRef fla) { return fla == logic.getTerm_true(); })) {
-                pmanager.assignTopLevelPartitionIndex(0, logic.getTerm_true());
-                status = giveToSolver(logic.getTerm_true(), frame.getId());
```

Comment:
Why did you exchange it for simple push on line 141?

Reply:
I am not sure what you mean. The line 141 in the new version does not correspond to this line.
It seems I have simple remove asserting literal `true` to the solver in case the frame is empty or consists of only `true` formulas. I am not sure why that was the case before. Either we can really just ignore such frames, or we are missing a test in our regression test suite.

Reply:
Hmm, ok, that is indeed weird
Well, in theory it shouldn't break anything, if it will, we will find out then 😅